server/src/main/java/com/thoughtworks/go/server/service/ConsoleActivityMonitor.java

```diff
@@ -124,6 +139,59 @@ private long jobTerminationThreshold(JobIdentifier jobIdentifier) {
         return goConfigService.getUnresponsiveJobTerminationThreshold(jobIdentifier);
     }
 
+    private interface LogMessages {
+        String consoleMessage(String difference);
+
+        String hungWarningMessage(String buildLocator, String namespacedJob, String difference);
+    }
+
+    private LogMessages scheduledJobMessages() {
+        return new LogMessages() {
+            @Override
+            public String consoleMessage(String difference) {
+                return format("Go cancelled this job as it was not assigned an agent for more than %s minute(s)", difference);
+            }
+
+            @Override
+            public String hungWarningMessage(String buildLocator, String namespacedJob, String difference) {
+                return format("Job <a href='/go/tab/build/detail/%s'>%s</a> is currently running but was not assigned an agent in the last %s minute(s). This job may be hung.", buildLocator, namespacedJob, difference);
+            }
+        };
+    }
+
+    private LogMessages runningJobMessages() {
+        return new LogMessages() {
+            @Override
+            public String consoleMessage(String difference) {
+                return format("Go cancelled this job as it has not shown any console activity for more than %s minute(s)", difference);
+            }
+
+            @Override
+            public String hungWarningMessage(String buildLocator, String namespacedJob, String difference) {
+                return format("Job <a href='/go/tab/build/detail/%s'>%s</a> is currently running but has not shown any console activity in the last %s minute(s). This job may be hung.", buildLocator, namespacedJob, difference);
+            }
+        };
+    }
+
+    static final class ScheduledJobListener implements JobStatusListener {
+        private final ConsoleActivityMonitor consoleActivityMonitor;
+
+        private ScheduledJobListener(ConsoleActivityMonitor consoleActivityMonitor) {
+            this.consoleActivityMonitor = consoleActivityMonitor;
+        }
+
+        @Override
+        public void jobStatusChanged(JobInstance job) {
+            JobIdentifier identifier = job.getIdentifier();
+            if (job.getState().isScheduled()) {
+                consoleActivityMonitor.jobScheduledMap.putIfAbsent(identifier, consoleActivityMonitor.timeProvider.currentTimeMillis());
+            } else if (job.getState().isActiveOnAgent() || job.isCompleted()) {
+                consoleActivityMonitor.jobScheduledMap.remove(identifier);
+                consoleActivityMonitor.removeHungJobWarning(identifier);
```

Comment:
Should we remove warning only if the `jobScheduledMap` has the identifier. Just wanted to make sure we do not clear out warnings added through `jobLastActivityMap`

Reply:
A job will be first added to `jobScheduledMap` when it gets scheduled. As soon as it gets assigned it will be removed from the map. As it starts building it will be added to `jobLastActivityMap`.
Hence, the same job identifier cannot be present in both the maps at the same time. 
Therefore, I think this is a safe operation.

Reply:
I get your point about how the jobs are added to the map.

Since the `ScheduledJobListener#jobStatusChanged` would be called for jobs which are part of `jobLastActivityMap`, was wondering if there will be cases where we end up clearing warnings for jobs which are in building state in this listener.